pkg/networkservice/common/refresh/client.go

```diff
@@ -55,76 +54,67 @@ func (t *refreshClient) Request(ctx context.Context, request *networkservice.Net
+	refreshRequest := request.Clone()
+	refreshRequest.GetConnection().Context = rv.Context
+	refreshRequest.GetConnection().Mechanism = rv.Mechanism
+	refreshRequest.GetConnection().NetworkServiceEndpointName = rv.NetworkServiceEndpointName
+
+	// TODO - introduce random noise into duration avoid timer lock
+	duration := time.Until(expireTime) / 3
 	t.executor.AsyncExec(func() {
-		id := req.GetConnection().GetId()
-		// check if it is refresh request
-		if refreshCtx := refreshContext(ctx); refreshCtx != nil {
-			// refresh was canceled
-			if refreshCtx.Err() != nil {
-				return
-			}
-			// we reuse non-canceled refresh context for the next refresh request
-			timer := t.createTimer(ctx, req, expire, opts...)
-			t.connectionTimers[id] = timer
-		} else {
-			// cancel refresh of previous request if any
-			if timer, ok := t.connectionTimers[id]; ok {
-				timer.Stop()
-			}
-			if canceller, ok := t.refreshCancellers[id]; ok {
-				canceller()
-			}
-			// add refresh context to request context
-			refreshCtx, cancelFunc := context.WithCancel(context.Background())
-			newCtx := withRefreshContext(ctx, refreshCtx)
+		// Create the refresh context
+		var cancel context.CancelFunc
+		refreshCtx := extend.WithValuesFromContext(t.ctx, ctx)
+		if deadline, ok := ctx.Deadline(); ok {
+			refreshCtx, cancel = context.WithDeadline(refreshCtx, deadline.Add(duration))
+		}
 
-			timer := t.createTimer(newCtx, req, expire, opts...)
-			t.connectionTimers[id] = timer
-			t.refreshCancellers[id] = cancelFunc
+		// Stop any existing timers
+		if timer, ok := t.timers[request.GetConnection().GetId()]; ok {
+			timer.Stop()
 		}
+
+		// Set new timer
+		var timer *time.Timer
+		timer = time.AfterFunc(duration, func() {
+			t.executor.AsyncExec(func() {
+				// Check to see if we've been superseded by another timer, if so, do nothing
+				currentTimer, ok := t.timers[refreshRequest.GetConnection().GetId()]
+				if ok && currentTimer != timer {
+					cancel()
+					return
+				}
+				// Resend request if the refreshCtx is not Done.  Note: refreshCtx should only be done if
+				// t.ctx has been canceled (usually because the clientCtx has been canceled
+				select {
+				case <-refreshCtx.Done():
+				default:
+					if _, err := t.Request(refreshCtx, refreshRequest, opts...); err != nil {
+						// TODO - do we want to retry at 2/3 and 3/3 if we fail here?
+						trace.Log(refreshCtx).Errorf("Error while attempting to refresh connection %s: %+v", request.GetConnection().GetId(), err)
+					}
+				}
+				// Set timer to nil to be really really sure we don't have a circular reference that precludes garbage collection
+				timer = nil
+			})
+		})
+		t.timers[request.GetConnection().GetId()] = timer
 	})
 	return rv, nil
 }
 
 func (t *refreshClient) Close(ctx context.Context, conn *networkservice.Connection, _ ...grpc.CallOption) (*empty.Empty, error) {
 	t.executor.AsyncExec(func() {
-		if timer, ok := t.connectionTimers[conn.GetId()]; ok {
```

Comment:
@edwarnicke I think we need to call cancel for refreshCtx, for more fast stop of Requesting in case of Close. 

Also possible we need to check on Close if Request is already in progress and wait until it will be canceled. Thoughts?

Reply:
If we weren’t in the executor you’d be right.  But because the executor guarantees exclusivity and order of execution, either 

1.  This invocation in Close comes first, in which case the timer is stopped and the timers invocation of executor never happens
2. The timer’s execution of executor happens first *and it completes* before we get to the Closes executor invocation here.  

Either way, executing cancel has no effect.

Reply:
No wait, you are correct, we could race executor is called, and the timer’s use could get in line while we are Closing.  Will fix.

Reply:
Fixed.